src/ppxgboost/PPPrediction.py

```diff
@@ -0,0 +1,129 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+import numpy as np
+import pandas as pd
+
+import ppxgboost.PaillierAPI as paillier
+from ppxgboost.PPModel import *
+from ppxgboost.PPQuery import PPQuery
+
+
+def predict_single_input_binary(model: PPModel, query, default_base_score=0.5):
+    """
+    Return the score for this query in a binary classification model
+
+    :param model: the model to evaluate
+    :param query: a single input query
+    :param default_base_score: a default score is 0.5 (global bias)
+    :return: the query's score
+    """
+
+    scores = model.eval(query)
+    return default_base_score + sum(scores)
+
+def predict_binary(enc_model: PPModel, enc_queries: [PPQuery], default_base_score=0.5):
+    """
+    Prediction on @enc_queries over the @enc_model
+    :param enc_model: encrypted PPModel
+    :param enc_queries: a list of encrypted PPQuery
+    :param default_base_score: default score is 0.5 (according to the xgboost -- global bias)
+    :return: a list of prediction values
+    """
+
+    return list(map(lambda q: predict_single_input_binary(enc_model, q, default_base_score), enc_queries))
+
+
+def predict_single_input_multiclass(model: PPModel, num_classes, query):
+    """
+    Return a prediction on a single input vector.
+    The algorithm computes the sum of scores for all the corresponding classes (boosters in the xgboost model).
+    For each class, it sum up all the leaves' values
+
+    :param model: internal model representation
+    :param num_classes: the total number classes to classify
+    :param query: a single input query
+    :return: the predicted score
+    """
+    scores = model.eval(query)
+
+    # sum of score for each category: exp(score)
+    result = []
```

Comment:
This looks like it could be greatly simplified with a Counter: https://docs.python.org/3/library/collections.html#collections.Counter